src/server/scripts/Shadowlands/SepulcherOfTheFirstOnes/boss_anduin_wrynn.cpp

```diff
@@ -0,0 +1,4307 @@
+    {
+        if (me->HealthAbovePctHealed(100, heal))
+        {
+            _instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+            DoCastSelf(SPELL_WILLPOWER_ENERGIZE_SMALL);
+            DoCastSelf(SPELL_SOUL_DESPAWN);
+
+            if (Creature* soul = me->GetInstanceScript()->GetCreature(DATA_ANDUIN_SOUL))
+                soul->GetAI()->DoAction(ACTION_HOPE_RESTORED);
+        }
+    }
+
+    void MovementInform(uint32 type, uint32 pointId) override
+    {
+        if (type != POINT_MOTION_TYPE)
+            return;
+
+        switch (pointId)
+        {
+            case POINT_ESCAPE_PLATFORM:
+            {
+                DoCastSelf(SPELL_SOUL_DESPAWN);
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+private:
+    InstanceScript* _instance;
+    TaskScheduler _scheduler;
+};
+
+// Fiendish Soul - 183669
+struct npc_anduin_wrynn_fiendish_soul : public ScriptedAI
+{
+    npc_anduin_wrynn_fiendish_soul(Creature* creature) : ScriptedAI(creature) { }
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_GHOUL_LEAP, 1s);
+        _events.ScheduleEvent(EVENT_NECROTIC_CLAWS, 10s);
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        if (IsHeroic() || IsMythic())
+            DoCastSelf(SPELL_SOUL_EXPLOSION_TARGET);
+
+        DoCastSelf(SPELL_SOUL_DESPAWN);
+    }
+
+    void EnterEvadeMode(EvadeReason /*why*/) override
+    {
+        me->DespawnOrUnsummon();
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
+
+        _events.Update(diff);
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_GHOUL_LEAP:
+                {
+                    std::list<Player*> targetList;
+                    GetPlayerListInGrid(targetList, me, 50.0f);
+                    me->CastSpell(Trinity::Containers::SelectRandomContainerElement(targetList), SPELL_NECROTIC_CLAWS_LEAP, false);
+                    _events.Repeat(10s);
+                    break;
+                }
+                case EVENT_NECROTIC_CLAWS:
+                {
+                    DoCastVictim(SPELL_NECROTIC_CLAWS_DEBUFF);
+                    _events.Repeat(5s, 10s);
+                    break;
+                }
+                default:
+                    break;
+            }
+        }
+
+        DoMeleeAttackIfReady();
+    }
+
+private:
+    EventMap _events;
+};
+
+// Monstrous Soul - 183671
+struct npc_anduin_wrynn_monstrous_soul : public ScriptedAI
+{
+    npc_anduin_wrynn_monstrous_soul(Creature* creature) : ScriptedAI(creature) { }
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        me->GetInstanceScript()->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, me, 255);
```

Comment:
we're calling GetInstanceScript without checking its nullance. save InstanceScript into a variable that we can call, since it's used multiple times in this struct.

Reply:
```cpp
template <class AI, class T>
inline AI* GetSepulcherOfTheFirstOnesAI(T* obj)
{
    return GetInstanceAI<AI>(obj, SFOScriptName);
}

#define RegisterSepulcherOfTheFirstOnesCreatureAI(ai_name) RegisterCreatureAIWithFactory(ai_name, GetSepulcherOfTheFirstOnesAI)

```
never null, script is not loaded outside instance